### PR TITLE
Performance optimization for isSymbolicLink() calls on Windows

### PR DESCRIPTION
**Pull request by AlexanderGolovlev**
Current implementation for WindowsFileSystem.isSymbolicLink() makes unnecessary system calls which significantly affect the performance. Original code goes through the following:
- AbstractFileSystemWithCustomStat.isSymbolicLink
  - WindowsFileSystem.stat
    - WindowsFileSystem.getIoFile
    - JavaIoFileSystem.getNioPath
    - Files.readAttributes
    - WindowsFileSystem.fileIsSymbolicLink
    - WindowsFileOperations.getLastChangeTime
This implementation skips most of them.

### CODE REVIEW

**Comment by fmeum** [comment]
Nice find!

**Review by tjgq — COMMENTED** [review]
The whole point of `AbstractFileSystemWithCustomStat` is to derive the `isFoo` methods from a `stat` implementation, so while this PR does seem worthwhile for the performance benefits, the end result seems a bit convoluted. I wonder if we could instead do the following:

- Have `JavaIoFileSystem` extend `AbstractFileSystem` instead of `AbstractFileSystemWithCustomStat`, with the `isFoo` methods implemented as `Files.isFoo`
- in `WindowsFileSystem`, override `isSymbolicLink` as currently implemented by your PR

But feel free to push back if my suggested refactoring is more difficult than it appears at first sight.

**Comment by AlexanderGolovlev** [comment]
@tjgq , with your comment I've got an idea that we could try to optimize other `isFoo` methods derived from `stat`, in addition to `isSymbolicLink`. This is done in my second commit and is based on a "lazy" implementation for some `stat` methods. This should save us at least 1/3 of time for any `isFoo` call, and 2/3 of time for some methods like `getFileSize`.
However, I'm not sure about replacing the `AbstractFileSystemWithCustomStat` by `AbstractFileSystem`. This will require implementing the `isFoo` methods not only in `JavaIoFileSystem`, but also in `WindowsFileSystem`. This will cause duplication of code and excessive complexity.

File: src/main/java/com/google/devtools/build/lib/windows/WindowsFileSystem.java [hunk 1]
```diff
@@ -148,31 +148,32 @@ protected FileStatus stat(PathFragment path, boolean followSymlinks) throws IOEx
       throw new FileNotFoundException(path + ERR_NO_SUCH_FILE_OR_DIR);
     }
 
-    final boolean isSymbolicLink = !followSymlinks && fileIsSymbolicLink(file);
-    final long lastChangeTime =
-        WindowsFileOperations.getLastChangeTime(getNioPath(path).toString(), followSymlinks);
     FileStatus status =
         new FileStatus() {
+          volatile Boolean isSymbolicLink;
```
Comment (tjgq) [inline src/main/java/com/google/devtools/build/lib/windows/WindowsFileSystem.java:153]: Let's mark this `@Nullable` and append a comment `// null if not yet known` to make the convention clear.

File: src/main/java/com/google/devtools/build/lib/windows/WindowsFileSystem.java [hunk 2]
```diff
@@ -148,31 +148,32 @@ protected FileStatus stat(PathFragment path, boolean followSymlinks) throws IOEx
...
 
           @Override
           public boolean isDirectory() {
-            return !isSymbolicLink && attributes.isDirectory();
+            return !isSymbolicLink() && attributes.isDirectory();
           }
 
           @Override
           public boolean isSymbolicLink() {
+            if (isSymbolicLink == null) { isSymbolicLink = !followSymlinks && fileIsSymbolicLink(file); }
```
Comment (tjgq) [inline src/main/java/com/google/devtools/build/lib/windows/WindowsFileSystem.java:176]: Nit: the `if` body should be on its own line.

File: src/main/java/com/google/devtools/build/lib/windows/WindowsFileSystem.java [hunk 3]
```diff
@@ -187,7 +188,8 @@ public long getLastModifiedTime() {
           }
 
           @Override
-          public long getLastChangeTime() {
+          public long getLastChangeTime() throws IOException {
+            if (lastChangeTime == -1) { lastChangeTime = WindowsFileOperations.getLastChangeTime(getNioPath(path).toString(), followSymlinks); }
```
Comment (tjgq) [inline src/main/java/com/google/devtools/build/lib/windows/WindowsFileSystem.java:192]: Same as above.

**Review by tjgq — CHANGES_REQUESTED** [review]
I like this version! Just a couple of minor comments.

**Comment by tjgq** [comment]
> @tjgq , with your comment I've got an idea that we could try to optimize other `isFoo` methods derived from `stat`, in addition to `isSymbolicLink`. This is done in my second commit and is based on a "lazy" implementation for some `stat` methods. This should save us at least 1/3 of time for any `isFoo` call, and 2/3 of time for some methods like `getFileSize`. However, I'm not sure about replacing the `AbstractFileSystemWithCustomStat` by `AbstractFileSystem`. This will require implementing the `isFoo` methods not only in `JavaIoFileSystem`, but also in `WindowsFileSystem`. This will cause duplication of code and excessive complexity.

That's a fair point; let's keep it an `AbstractFileSystemWithCustomStat`, then.

**Review by tjgq — APPROVED** [review]

**Pull request closed without merging**